src/IO/ReadWriteBufferFromHTTP.h

```diff
@@ -277,17 +293,45 @@ namespace detail
                                 settings.http_max_tries, settings.http_retry_initial_backoff_ms, settings.http_retry_max_backoff_ms);
 
             if (!delay_initialization)
+            {
                 initialize();
+                if (exception)
+                    std::rethrow_exception(exception);
+            }
+        }
+
+        void call(Poco::Net::HTTPResponse & response, const String & method_)
+        {
+            try
+            {
+                istr = callImpl(saved_uri_redirect ? *saved_uri_redirect : uri, response, method_);
+            }
+            catch (...)
+            {
+                if (response.getStatus() == Poco::Net::HTTPResponse::HTTPStatus::HTTP_NOT_FOUND
+                    && http_skip_not_found_url)
+                {
+                    initialization_error = InitializeError::SKIP_NOT_FOUND_URL;
+                }
+                else
+                {
+                    throw;
+                }
+            }
         }
 
         /**
-         * Note: In case of error return false if error is not retriable, otherwise throw.
+         * Throws if error is not retriable, otherwise sets initialization_error = NON_RETRIABLE_ERROR and
+         * saves exception into `exception` variable. In case url is not found and skip_not_found_url == true,
+         * sets initialization_error = DKIP_NOT_FOUND_URL, otherwise throws.
```

Comment:
```suggestion
         * Throws if error is retriable, otherwise sets initialization_error = NON_RETRIABLE_ERROR and
         * saves exception into `exception` variable. In case url is not found and skip_not_found_url == true,
         * sets initialization_error = SKIP_NOT_FOUND_URL, otherwise throws.
```